source/azure_iot_hub_client.c

```diff
@@ -626,3 +724,503 @@ AzureIoTHubClientResult_t AzureIoTHubClient_SendTelemetry( AzureIoTHubClient_t *
 
     return xResult;
 }
+/*-----------------------------------------------------------*/
+
+AzureIoTHubClientResult_t AzureIoTHubClient_ProcessLoop( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                         uint32_t ulTimeoutMilliseconds )
+{
+    AzureIoTMQTTResult_t xMQTTResult;
+    AzureIoTHubClientResult_t xReturn;
+
+    if( pxAzureIoTHubClient == NULL )
+    {
+        AZLogError( ( "Failed to process loop: Invalid argument" ) );
+        xReturn = eAzureIoTHubClientInvalidArgument;
+    }
+    else if( ( xMQTTResult = AzureIoTMQTT_ProcessLoop( &( pxAzureIoTHubClient->_internal.xMQTTContext ),
+                                                       ulTimeoutMilliseconds ) ) != eAzureIoTMQTTSuccess )
+    {
+        AZLogError( ( "AzureIoTMQTT_ProcessLoop failed : ProcessLoopDuration=%u, Error=0x%08x",
+                      ulTimeoutMilliseconds, xMQTTResult ) );
+        xReturn = eAzureIoTHubClientFailed;
+    }
+    else
+    {
+        xReturn = eAzureIoTHubClientSuccess;
+    }
+
+    return xReturn;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTHubClientResult_t AzureIoTHubClient_SubscribeCloudToDeviceMessage( AzureIoTHubClient_t * pxAzureIoTHubClient,
```

Comment:
I want to talk through the disconnect scenarios esp. with SUBSCRIBE.  I think code we have is fine but let's make sure my mental model is correct.

So we do Connect/SubscribeC2D (e.g.).  And then we disconnect.  So nothing's happening.

Then assume app does a AzIoT_Connect again and it succeeds.

If App is using session present and hub accepts, then nothing has NULL'd out `pxContext->_internal.pxProcessFunction` so we're still in business to deal with callbacks.

If app isn't using session present OR hub accepts connect but not new session, then app would be on its own to re-subscribe which would reset all these guys.

I know this is a long comment indicating I'm happy, but if I'm misreading this I need to be educated.  These retry issues can be bug farms.

Reply:
oh interesting. So essentially what is the behavior of cloud to device if we have a disconnect then connect again with session present?

Reply:
Yes, correct. Client when connect need to inspect session present or not (assuming clean session is pass false). If session is present on server, client need to set its state. Client can set its state if not set previously (e.g If done init on iothub client, client need to subscribe topic to set its state, otherwise no subscribe is required )

Reply:
should we wipe the receive contexts on a connect?

Reply:
no, in cases when user re-connect with clean session false, he should be allowed to receive messages without subscribe call.

---
In reply to: [612833270](https://github.com/Azure/azure-iot-middleware-freertos/pull/75#discussion_r612833270) [](ancestors = 612833270)